client/components/util/work-queue-service.js

```diff
@@ -0,0 +1,108 @@
+/**
+ * @copyright Copyright 2015 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * @fileoverview workQueue is an angular service that manages a queue
+ * of pending work item promises, running them with concurrency and
+ * priority control.
+ * @author klausw@google.com (Klaus Weidner)
+ */
+
+goog.provide('p3rf.perfkit.explorer.components.util.WorkQueueService');
+
+goog.scope(function() {
+const explorer = p3rf.perfkit.explorer;
+
+/**
+ * See module docstring for more information about purpose and usage.
+ *
+ * @param {!angular.$q} $q
+ * @param {angular.$timeout} $timeout
+ * @constructor
+ * @ngInject
+ */
+explorer.components.util.WorkQueueService = function($q, $timeout) {
+  this.q_ = $q;
+  this.timeout_ = $timeout;
+  this.queuePending_ = [];
+  this.queueExecuting_ = 0;
+};
+const WorkQueueService = explorer.components.util.WorkQueueService;
+
+WorkQueueService.NOTIFICATION = {
+  STARTED: 'Started'
+};
+
+WorkQueueService.MAX_PARALLEL_QUERIES_DEFAULT = 4;
```

Comment:
This should be a service level setting, stored in [DataStore](https://github.com/GoogleCloudPlatform/PerfKitExplorer/blob/master/server/perfkit/explorer/model/explorer_config.py#L45), and exposed in [ConfigService](https://github.com/GoogleCloudPlatform/PerfKitExplorer/blob/master/client/components/config/config-service.js#L44) and the [config](https://github.com/GoogleCloudPlatform/PerfKitExplorer/blob/master/client/components/config/config-directive.html) dialog directive.